### PR TITLE
updates vv

### PR DESCRIPTION
**Pull request by harryob**
CleanShot 2023-01-20 at 12 35 23@2x
CleanShot 2023-01-20 at 12 36 33@2x

:cl:
refactor: refactor's admin VV
admin: can now edit lists with any reliability, can now edit bitfields without maths
ui: added the particle editor and colour matrix editor (from tiviplus and Y0SH1M4S73R respectively) for admins via VV
/:cl:

### CODE REVIEW

File: code/__DEFINES/generators.dm [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+#define GEN_NUM "num"
+#define GEN_VECTOR "vector"
+#define GEN_BOX "box"
+#define GEN_COLOR "color"
+#define GEN_CIRCLE "circle"
+#define GEN_SPHERE "sphere"
+#define GEN_SQUARE "square"
+#define GEN_CUBE "cube"
+
+///particle editor var modifiers
```
Comment (fira) [inline code/__DEFINES/generators.dm:11]:
```suggestion
//particle editor var modifiers
```

File: code/__DEFINES/traits.dm [hunk 2]
```diff
@@ -210,6 +210,70 @@ GLOBAL_LIST_INIT(mob_traits, list(
...
+
+/// value -> trait name, generated on use from trait_by_type global
+GLOBAL_LIST(trait_name_map)
+
+/proc/generate_trait_name_map()
+	. = list()
+	for(var/key in GLOB.traits_by_type)
+		for(var/tname in GLOB.traits_by_type[key])
+			var/val = GLOB.traits_by_type[key][tname]
+			.[val] = tname
```
Comment (fira) [inline code/__DEFINES/traits.dm:273]: but, this is never used nor returned?

File: code/game/world.dm [hunk 3]
```diff
@@ -83,7 +83,9 @@ var/list/reboot_sfx = file2list("config/reboot_sfx.txt")
 
 	if(!EvacuationAuthority) EvacuationAuthority = new
 
+	world.log << world.tick_lag
 	change_tick_lag(CONFIG_GET(number/ticklag))
+	world.log << CONFIG_GET(number/ticklag)
```
Comment (fira) [inline code/game/world.dm:88]: leftover debug?

File: code/modules/admin/view_variables/filterrific.dm [hunk 4]
```diff
@@ -78,7 +78,8 @@
 				target.update_filters()
 				. = TRUE
 		if("mass_apply")
-			if(!check_rights_for(usr.client, R_DEBUG))
+			if(!check_rights_for(usr.client, R_EVENT))
+				to_chat(usr, SPAN_USERDANGER("Stay in your lane, jannie."))
```
Comment (fira) [inline code/modules/admin/view_variables/filterrific.dm:82]: message aside, is the change from R_DEBUG to R_EVENT intended ? might need to bring this up with server management

File: code/modules/admin/view_variables/reference_tracking.dm [hunk 5]
```diff
@@ -0,0 +1,161 @@
+#ifdef REFERENCE_TRACKING
```
Comment (fira) [inline code/modules/admin/view_variables/reference_tracking.dm:1]: did you give ref tracking a test ? i don't remember for sure if it was fully added in on SSgarbage during backport. either way it's good to have in the future

File: code/modules/admin/view_variables/view_variables.dm [hunk 6]
```diff
@@ -0,0 +1,279 @@
+/client/proc/debug_variables(datum/D in world)
+	set category = "Debug"
+	set name = "View Variables"
+	//set src in world
+	var/static/cookieoffset = rand(1, 9999) //to force cookies to reset after the round.
+
+	if(!usr.client || !usr.client.admin_holder) //This is usr because admins can call the proc on other clients, even if they're not admins, to show them VVs.
```
Comment (fira) [inline code/modules/admin/view_variables/view_variables.dm:7]: might need to adapt such checks due to mentors

File: code/modules/admin/view_variables/filterrific.dm (hunk 4 above)
Comment (harryob) [reply code/modules/admin/view_variables/filterrific.dm:82]: nope, it (with the message) was just upstream. both nuked

File: code/game/world.dm (hunk 3 above)
Comment (harryob) [reply code/game/world.dm:88]: huh

File: code/modules/admin/view_variables/reference_tracking.dm (hunk 5 above)
Comment (harryob) [reply code/modules/admin/view_variables/reference_tracking.dm:1]: it didn't - now it does!

**Review by fira — APPROVED** [review]

**Review by Nanu308 — CHANGES_REQUESTED** [review]
Remove decrepit tools/verbs from the VV drop down menu and clean out some of that code.

Ex. Delimb, Relimb, Amputate etc.

**Comment by harryob** [comment]
> Remove decrepit tools/verbs from the VV drop down menu and clean out some of that code.
> 
> Ex. Delimb, Relimb, Amputate etc.

done

**Review by stanalbatross — APPROVED** [review]
lgtm + fira approved

**Review by Nanu308 — APPROVED** [review]

**Pull request merged**